### PR TITLE
check subprocess.returncode to print error message.

### PR DESCRIPTION
**Pull request by fujitatomoya**
address https://github.com/ros2/ros2cli/issues/601

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

File: ros2run/ros2run/api/__init__.py [hunk 1]
```diff
@@ -66,6 +67,8 @@ def run_executable(*, path, argv, prefix=None):
             # the subprocess will also receive the signal and should shut down
             # therefore we continue here until the process has finished
             pass
+    if process.returncode < 0:
+        print(signal.strsignal(-process.returncode))
```
Comment (fujitatomoya) [inline ros2run/ros2run/api/__init__.py:71]: the alternative is we can wrap subprocess with `shell=True` but i just do not like to instantiate shell process only for this.

**Comment by fujitatomoya** [comment]
@audrow could you take a look when you got time?

File: ros2run/ros2run/api/__init__.py (hunk 1 above)
Comment (clalancette) [reply ros2run/ros2run/api/__init__.py:71]: I don't think we should use `shell=True` either.

As far as the code here, can `returncode` ever be `< 0`?  In Linux, at least, return codes are always 0 or greater.

Comment (fujitatomoya) [reply ros2run/ros2run/api/__init__.py:71]: yeah, it is python spec which you can see https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.returncode.

`returncode` will be negative if the child process is terminated with signal. (e.d `Segmentation Fault` will be `-11`)

Comment (clalancette) [reply ros2run/ros2run/api/__init__.py:71]: > yeah, it is python spec which you can see https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.returncode.

Oh!  I didn't realize that, thanks for pointing it out.

The documentation there says:

```
A negative value -N indicates that the child was terminated by signal N (POSIX only).
```

So I wonder what happens on Windows.

**Review by audrow — APPROVED** [review]
Looks like a good change to me. Is there a way to add a test for it?

File: ros2run/ros2run/api/__init__.py (hunk 1 above)
Comment (fujitatomoya) [reply ros2run/ros2run/api/__init__.py:71]: that is actually nice question, i am not really sure. (dont have environment.) does anyone happen to know about this?
CC: @iuhilnehc-ynos @Barry-Xu-2018

Comment (Barry-Xu-2018) [reply ros2run/ros2run/api/__init__.py:71]: @fujitatomoya  @clalancette 

I check return code on window.  
Run ` ros2 run demo_nodes_cpp talker` and terminate `talker.exe` in windows task manager.  
The return code is `1`.   

According to wiki [exit status for window](https://en.wikipedia.org/wiki/Exit_status#Windows), it provides a reference to [Windows System Error Codes (exit codes)](http://www.hiteksoftware.com/knowledge/articles/049.htm). In this document, exit codes are  all  greater than 0 if error occurs or terminated by other process.

Comment (fujitatomoya) [reply ros2run/ros2run/api/__init__.py:71]: thanks @Barry-Xu-2018 

then we can do the following? any suggestions are welcome.

1. check if returcode is NOT zero?
2. if POSIX, call strsignal. and else print general message that subprocess exited with failure.
3. return `returncode`

Comment (clalancette) [reply ros2run/ros2run/api/__init__.py:71]: I also did a bit of testing on Windows.  I wrote this silly code:

```
#include <iostream>

int main()
{
    std::cout << "Hello, world" << std::endl;
    *(char *)0 = 0;
    return 0;
}
```

Which obviously causes a crash.  If I compile and run that program on the command-line, I get a negative error code:

```
C:\dev>cl /EHsc crash.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29336 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

crash.cpp
Microsoft (R) Incremental Linker Version 14.28.29336.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:crash.exe
crash.obj

C:\dev>crash
Hello, world

C:\dev>echo %errorlevel%
-1073741819
```

However, I get a different result if I run it through subprocess on Python.  Here is my simple Python code:

```
import subprocess

process = subprocess.Popen(['crash'])
process.communicate()
print(process.returncode)
```

When I run that, I get:

```
C:\dev>python subprocess-crash.py
Hello, world
3221225477

```

So it looks like Python is wrapping the negative return code or something like that.

Regardless, I think @fujitatomoya 's proposal sounds like an improvement over the current situation, so :+1: from me.

**Comment by fujitatomoya** [comment]
@clalancette i'd like to request a review.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14703/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9455/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12368/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14847/)

File: ros2run/ros2run/api/__init__.py [hunk 2]
```diff
@@ -66,6 +67,13 @@ def run_executable(*, path, argv, prefix=None):
             # the subprocess will also receive the signal and should shut down
             # therefore we continue here until the process has finished
             pass
+    if process.returncode != 0:
+        if os.name == 'posix':
+            # a negative value -N indicates that the child was terminated by signal N.
+            print(signal.strsignal(-process.returncode))
```
Comment (clalancette) [inline ros2run/ros2run/api/__init__.py:73]: If the process doesn't crash, but returns a non-zero exit code, this ends up with:

```
ValueError: signal number out of range
```

So we'll have to handle that differently.

File: ros2run/ros2run/api/__init__.py [hunk 3]
```diff
@@ -66,6 +67,13 @@ def run_executable(*, path, argv, prefix=None):
...
+        else:
+            # print general failure message instead.
+            print('Process exits failure')
```
Comment (clalancette) [inline ros2run/ros2run/api/__init__.py:76]:
```suggestion
            print('Process exited with failure %d' % (process.returncode))
```

**Review by clalancette — CHANGES_REQUESTED** [review]
Besides the comments inline, I'm thinking we should have some prefix to show that these messages came from `ros2run`, and not from the process itself.  Maybe something like "[ros2run]: Message".

File: ros2run/ros2run/api/__init__.py (hunk 2 above)
Comment (fujitatomoya) [reply ros2run/ros2run/api/__init__.py:73]: Ah i see, nice catch 👍 thanks. i will do check valid signal range before that.

**Comment by fujitatomoya** [comment]
@clalancette all comments are addressed, i will run CI.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14716/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9468/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12383/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14861/)

**Comment by fujitatomoya** [comment]
@clalancette friendly ping.

**Comment by clalancette** [comment]
> Besides the comments inline, I'm thinking we should have some prefix to show that these messages came from `ros2run`, and not from the process itself. Maybe something like "[ros2run]: Message".

This comment is still open.  I do think we should have some kind of prefix on the output so people can tell it comes from `ros2run` and not from their binary.

**Comment by fujitatomoya** [comment]
@clalancette my bad, i missed that. i will address it.

**Comment by fujitatomoya** [comment]
unit test to confirm the message. (using simple example to generate segmentation fault)

https://github.com/fujitatomoya/ros2_test_prover/blob/master/prover_rclcpp/src/ros2cli_601.cpp

```
root@134f29e8f25f:~/ros2_ws/colcon_ws# ros2 run prover_rclcpp ros2cli_601
Doing something terrible!
[ros2run]: Segmentation fault
```

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14738/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9489/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12403/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14884/)

**Review by clalancette — APPROVED** [review]

**Comment by audrow** [comment]
Rerunning CI (just in case):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14770/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9519/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12436/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14917/)

**Comment by audrow** [comment]
Thanks for the PR, @fujitatomoya!

**Pull request merged**